Show flag summary information to group moderators

fatbusinessman opened the pull request:
When presenting annotations for someone who is a moderator in their group, this will now include a `moderation` key in the annotation JSON. Currently it just shows the number of times each annotation has been flagged for the moderator’s attention.

My current plan:

- [x] Pull out the model logic into a service, akin to #4475 – this could go in a subsequent PR
- [x] Give the commits some hefty rejigging
- [x] Consider changing the name of the formatter class, since it could now be confused with the `AnnotationModeration` model
- [x] Decide whether to keep the functional test, remove it, or add to it

This work is for hypothesis/product-backlog#208.

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+            return {}
+
+        flag_count = self._load(annotation_resource.annotation.id)
+        return {'moderation': {'flag_count': flag_count}}
```
Small thing, but I think we'll probably want `flagCount` here. We don't have a lot of multiword keys in the API output, the one exception being the selectors for the annotation, but those use `lowerCamelCase` rather than `lower_underscored`.

nickstenning commented:
This is looking pretty good to me! FWIW, my feedback on your to-do list:

- services good ✔️ 
- hefty rejigging good ✔️ 
- meh, don't worry too much about the name -- we have package namespaces and [they're cool](https://www.python.org/dev/peps/pep-0020/)
- perhaps we could work the functional test into a more generic functional test of the annotation API in totality? (i.e. prepare a couple of annotations with various data, and check the output is as we expect, rather than checking each feature individually)

nickstenning commented:
Oh, and I think having to pass `request.has_permission` around is a little unfortunate, but I can't immediately think of a better idea, so... ¯\_(ツ)_/¯

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+
+    def preload(self, ids):
+        if self._user is None:
+            return
```
We probably also want an `if not ids` shortcut here, too.

fatbusinessman replied:
Good point; added.

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+        query = self._session.query(sa.func.count(models.Flag.id).label('flag_count'),
+                                    models.Flag.annotation_id) \
+                             .filter(models.Flag.annotation_id.in_(ids)) \
+                             .group_by(models.Flag.annotation_id)
```
This is very much a take it or leave it comment, but I tend to think that using parens for multi-line queries is a bit neater than escaping newlines:

```python
        query = (self._session.query(sa.func.count(models.Flag.id).label('flag_count'),
                                     models.Flag.annotation_id)
                 .filter(models.Flag.annotation_id.in_(ids))
                 .group_by(models.Flag.annotation_id))
```

fatbusinessman replied:
Good point. I personally prefer that too; I was following the convention I’ve seen elsewhere in the codebase, but I’d be happy to go with this one instead.

Given this is code I’m moving around in #4502, I’m going to take your suggestion, but over in the other pull request.

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+                             .group_by(models.Flag.annotation_id)
+
+        flag_counts = {f.annotation_id: f.flag_count for f in query}
+        missing_ids = set(ids) - set(flag_counts.keys())
```
Again, certainly not a blocker for merging this, but it might be slightly less work to do this the other way round. That is:

```python
flag_counts = {id_: 0 for id_ in ids}
flag_counts.update({f.annotation_id: f.flag_count for f in query})
```

fatbusinessman replied:
You’re right: it would. Again, I’ll address this over in #4502.

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+
+    def _load(self, id_):
+        if self._user is None:
+            return False
```
It would be good to add a test for the negative branch here.

fatbusinessman replied:
Having looked at it again, I realised that this path will never actually get run, because if the user is `None`, they’ll fail the permission check in `format` and this method won’t get called. I’ve removed the check.

nickstenning commented on h/formatters/annotation_moderation.py:
```diff
@@ -0,0 +1,67 @@
+            return False
+
+        if id_ in self._cache:
+            return self._cache[id_]
```
Ditto -- we should cover the positive branch here.

fatbusinessman replied:
Done.

nickstenning commented on tests/h/formatters/annotation_moderation_test.py:
```diff
@@ -0,0 +1,97 @@
+    def test_preload_sets_flag_counts(self, db_session, flagged, unflagged, user):
+        formatter = AnnotationModerationFormatter(db_session,
+                                                  user,
+                                                  has_permission=None)
```
`has_permission=None` is  presumably a convenience, in the sense that you just happen not to be calling any methods in this test that would explode because you haven't passed it in?

fatbusinessman replied:
Correct. I’ve managed to eliminate most of these cases (except the one where the user is `None`) by extracting a fixture for the common case (of a user who has permission to see the `moderation` key).

fatbusinessman commented:
Ok. I think I’ve addressed all the feedback (thanks, @nickstenning!) and rebased onto `master` to accommodate #4499. Give or take Jenkins having a bit of a sad at the moment, I hope this is ready for review again (and, give or take some `autosquash` magicks that I’ve deferred to make review easier, for merging).

nickstenning commented:
LGTM pending a `rebase --autosquash`.
